### PR TITLE
feat: add save/load to ChiSquared

### PR DESCRIPTION
**Pull request by lalo**

### CODE REVIEW

**Comment by lalo** [comment]
Originally from: #3252

File: vowpalwabbit/distributionally_robust.h [hunk 1]
```diff
@@ -29,6 +30,30 @@ namespace distributionally_robust
     {
     }
     double qfunc(double w, double r) { return unbounded ? 1 : -(gamma + (beta + r) * w) / ((n + 1) * kappa); }
+
+    void save_load_aml(io_buf& model_file, bool read, bool text)
```
Comment (jackgerrits) [inline vowpalwabbit/distributionally_robust.h:34]: Why aml? Is this different or just a regular save/load function for DRO?

File: vowpalwabbit/distributionally_robust.h [hunk 2]
```diff
@@ -118,6 +143,64 @@ namespace distributionally_robust
       return duals.second.qfunc(w, r);
     }
 
+    void save_load_aml(io_buf& model_file, bool read, bool text)
```
Comment (lalo) [inline vowpalwabbit/distributionally_robust.h:146]:
```suggestion
    void save_load(io_buf& model_file, bool read, bool text)
```

File: vowpalwabbit/distributionally_robust.h (hunk 1 above)
Comment (lalo) [reply vowpalwabbit/distributionally_robust.h:34]:
```suggestion
    void save_load(io_buf& model_file, bool read, bool text)
```

File: vowpalwabbit/distributionally_robust.h [hunk 3]
```diff
@@ -118,6 +143,64 @@ namespace distributionally_robust
...
+        std::stringstream msg;
+        if (!read) msg << "_aml_chi_scored " << duals.first << "\n";
+        bin_text_read_write_fixed(
+            model_file, reinterpret_cast<char*>(&duals.first), sizeof(duals.first), "", read, msg, text);
+
+        if (!read) msg << "_aml_chi_duals_stale " << duals_stale << "\n";
+        bin_text_read_write_fixed(
+            model_file, reinterpret_cast<char*>(&duals_stale), sizeof(duals_stale), "", read, msg, text);
+
+        duals.second.save_load_aml(model_file, read, text);
```
Comment (lalo) [inline vowpalwabbit/distributionally_robust.h:160]:
```suggestion
        duals.second.save_load(model_file, read, text);
```

File: vowpalwabbit/distributionally_robust.h [hunk 4]
```diff
@@ -29,6 +30,30 @@ namespace distributionally_robust
...
     }
     double qfunc(double w, double r) { return unbounded ? 1 : -(gamma + (beta + r) * w) / ((n + 1) * kappa); }
+
+    void save_load(io_buf& model_file, bool read, bool text)
+    {
+      if (model_file.num_files() == 0) { return; }
+      if (!read || true)
+      {
+        std::stringstream msg;
+        if (!read) msg << "_aml_duals_unbounded " << unbounded << "\n";
```
Comment (jackgerrits) [inline vowpalwabbit/distributionally_robust.h:40]: Please use braces for all if statements

File: vowpalwabbit/distributionally_robust.h [hunk 5]
```diff
@@ -29,6 +30,30 @@ namespace distributionally_robust
     {
     }
     double qfunc(double w, double r) { return unbounded ? 1 : -(gamma + (beta + r) * w) / ((n + 1) * kappa); }
+
+    void save_load(io_buf& model_file, bool read, bool text)
+    {
+      if (model_file.num_files() == 0) { return; }
+      if (!read || true)
```
Comment (jackgerrits) [inline vowpalwabbit/distributionally_robust.h:37]: This will always evaluate to true - why have an if statement?

Comment (lalo) [reply vowpalwabbit/distributionally_robust.h:37]: placeholder for model version but forgot to completely remove, good catch

**Review by jackgerrits — APPROVED** [review]

**Pull request merged**